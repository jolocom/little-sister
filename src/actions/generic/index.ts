import { AppError, ErrorCode } from '../../lib/errors'
import I18n from '../../locales/i18n'
import { routeList } from '../../routeList'
import { navigationActions } from '../../actions'
import { ThunkAction } from '../../store'
import settingKeys from '../../ui/settings/settingKeys'

import * as Keychain from 'react-native-keychain'
// TODO use the settings items from storage
import AsyncStorage from '@react-native-community/async-storage'
// TODO don't depend on the crypto lib, perhaps use the rust crypto utils?
import crypto from 'crypto'

import { termsOfServiceDE } from 'src/ui/termsAndPrivacy/legalTexts'

import {
  AppWrapConfig,
  APPWRAP_UPDATE_CONFIG,
  APPWRAP_SHOW_LOADER,
  APPWRAP_REGISTER_CONFIG,
  APPWRAP_UNREGISTER_CONFIG,
  APPWRAP_SET_LOCKED,
  APPWRAP_SET_DISABLE_LOCK,
} from '../../reducers/generic'
import { AnyAction } from 'redux'
import { PIN_SERVICE } from 'src/ui/deviceauth/utils/keychainConsts'

// Default delay on the loading state value before it can switch back to 'false'
const DEFAULT_LOADING_LATCH_DELAY_MS = 300

export const showErrorScreen = (
  error: AppError | Error,
): ThunkAction => dispatch => {
  const appError: AppError =
    error.constructor === AppError
      ? (error as AppError)
      : new AppError(ErrorCode.Unknown, error)

  return dispatch(
    navigationActions.navigate({
      routeName: routeList.Exception,
      params: {
        returnTo: appError.navigateTo,
        error: appError,
      },
    }),
  )
}

const hashString = (text: string) => {
  return crypto
    .createHash('sha256')
    .update(text)
    .digest('hex')
}

const TERMS_OF_CONDITIONS_KEY = 'TERMS_OF_CONDITIONS'
export const checkTermsOfService = (
  route: routeList,
  onSubmit?: () => void,
): ThunkAction => async dispatch => {
  const storageHash = await AsyncStorage.getItem(TERMS_OF_CONDITIONS_KEY)
  const currentHash = hashString(termsOfServiceDE)
  const shouldShowTerms = storageHash !== currentHash

  if (!shouldShowTerms && onSubmit) onSubmit()

  if (shouldShowTerms) {
    return dispatch(
      navigationActions.navigate({
        routeName: routeList.TermsOfServiceConsent,
        params: { nextRoute: route, onSubmit },
      })
    )
  } else {
    return dispatch(
      navigationActions.navigate({
        routeName: route,
      })
    )
  }
}

export const storeTermsOfService = (
  route: routeList,
): ThunkAction => async dispatch => {
  const termsHash = hashString(termsOfServiceDE)
  await AsyncStorage.setItem(TERMS_OF_CONDITIONS_KEY, termsHash)

  dispatch(navigationActions.navigate({ routeName: route }))
}

export const setLocale = (locale: string): ThunkAction => async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  await backendMiddleware.storage.store.setting(settingKeys.locale, {
    selected: locale
  })

  I18n.locale = locale
  dispatch({ type: 'SET_LOCALE', value: locale })

  // we need to reset the navigator so that all screens are re-rendered with the
  // new locale
  return dispatch(navigationActions.navigatorReset())
}

const setLoading = (value: boolean) => ({
  type: APPWRAP_SHOW_LOADER,
  value,
})

export const setLocked = (value: boolean) => ({
  type: APPWRAP_SET_LOCKED,
  value,
})

export const setDisableLock = (value: boolean) => ({
  type: APPWRAP_SET_DISABLE_LOCK,
  value,
})

export const lockApp = (
): ThunkAction => async (dispatch, getState) => {
  // if the lock is globally disabled, or there is no identity created/loaded
  // yet, then this action has no effect
  const disableLock = getState().generic.disableLock
  const hasIdentity = !!getState().account.did.did
  if (disableLock || !hasIdentity) return

  // otherwise we lock the app, using the appropriate lock screen
  const keychainPin = await Keychain.getGenericPassword({
    service: PIN_SERVICE,
  })
  const lockScreen = keychainPin ? routeList.Lock : routeList.RegisterPIN

  dispatch(setLocked(true))
  return dispatch(navigationActions.navigate({ routeName: lockScreen }))
}

export const unlockApp = (
  pin: string,
  resetNav?: boolean,
): ThunkAction => async dispatch => {
  const keychainPin = await Keychain.getGenericPassword({
    service: PIN_SERVICE,
  })

  if (!keychainPin || keychainPin.password !== pin) throw new Error('wrongPin')

  dispatch(setLocked(false))
  return resetNav
    ? dispatch(navigationActions.navigatorResetHome())
    : dispatch(navigationActions.navigateBack())
}

export const updateAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UPDATE_CONFIG,
  value,
})
export const registerAppWrapConfig = (value: AppWrapConfig & {}) => ({
  type: APPWRAP_REGISTER_CONFIG,
  value,
})
export const unregisterAppWrapConfig = (value: AppWrapConfig) => ({
  type: APPWRAP_UNREGISTER_CONFIG,
  value,
})

export const showAppLoading = (
  shouldShow: boolean,
  latchDelay = DEFAULT_LOADING_LATCH_DELAY_MS,
): ThunkAction => async (dispatch, getState) => {
  const wasShowing = getState().generic.appWrapConfig.loading
  if (shouldShow && !wasShowing) {
    // for setting to "true" we do it immediately and set current state
    const ret = dispatch(setLoading(shouldShow))
    return new Promise<AnyAction>(resolve => {
      setTimeout(() => resolve(ret), latchDelay)
    })
  }

  if (wasShowing && !shouldShow) {
    // for setting to 'false', we delay by the latchDelay amount
    return dispatch(delayedSetLoading(shouldShow, latchDelay))
  }
}

let nextLoadingState = false
let timeout: number | undefined
const delayedSetLoading = (
  value: boolean,
  latchDelay: number,
): ThunkAction => async dispatch => {
  nextLoadingState = value
  if (timeout !== undefined) clearTimeout(timeout)
  timeout = setTimeout(() => {
    dispatch(setLoading(nextLoadingState))
    timeout = undefined
  }, latchDelay)
}
