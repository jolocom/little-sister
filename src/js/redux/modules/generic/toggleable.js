import * as _ from 'lodash'
import { action } from '../'

export function genericReducer(actions, {initialState}) {
  return function reducer(state = initialState, action = {}) {
    switch (action.type) {
      case actions.show.id:
        return true
      case actions.hide.id:
        return false
      case actions.toggle.id:
        return !state
      default:
        return state
    }
  }
}

export default function creator(prefix, type, {initialValue = false}) {
  const camelCaseType = _.camelCase(type)
  const actions = {
    show: action(prefix, 'show' + camelCaseType, {
      expectedParams: []
    }),
    hide: action(prefix, 'hide' + camelCaseType, {
      expectedParams: []
    }),
    toggle: action(prefix, 'toggle' + camelCaseType, {
      expectedParams: []
    })
  }
  return {
    reducer: genericReducer(actions, {initialState: initialValue}),
    actions
  }
}
