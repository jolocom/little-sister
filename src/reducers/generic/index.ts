import { AnyAction } from 'redux'
import { pickBy } from 'ramda'

export const APPWRAP_UPDATE_CONFIG = 'APPWRAP_UPDATE_CONFIG'
export const APPWRAP_SHOW_LOADER = 'APPWRAP_SHOW_LOADER'
export const APPWRAP_REGISTER_CONFIG = 'APPWRAP_REGISTER_CONFIG'
export const APPWRAP_UNREGISTER_CONFIG = 'APPWRAP_UNREGISTER_CONFIG'
export const APPWRAP_SET_LOCKED = 'APPWRAP_SET_LOCKED'
export const APPWRAP_SET_DISABLE_LOCK = 'APPWRAP_SET_DISABLE_LOCK'

export interface AppWrapConfig {
  readonly withoutStatusBar: boolean
  readonly loading: boolean
  readonly dark: boolean
  readonly secondaryDark: boolean
}
const initialAppWrapAttrs: AppWrapConfig = {
  loading: false,
  withoutStatusBar: false,
  dark: false,
  secondaryDark: false,
}

export const isAppWrapConfigAttr = (key: string) => initialAppWrapAttrs[key] !== undefined
export const pickAppWrapConfigAttrs: (o: any) => AppWrapConfig =
  pickBy((v, k) => isAppWrapConfigAttr(k))

export interface AppWrapState {
  readonly locked: boolean
  readonly disableLock: boolean
  readonly appWrapConfig: AppWrapConfig
  readonly appWrapConfigsSet: AppWrapConfig[]
}
const initialAppWrapState: AppWrapState = {
  locked: false,
  disableLock: false,
  appWrapConfig: initialAppWrapAttrs,
  appWrapConfigsSet: []
}

// TODO(mnzaki): this code is unnecessarily complicated
// the complicated "ConfigsSet" keeping is to support nested <Wrapper>
// components
// - but we don't really need to support them
// - see 20APR[RN61.5] on notion.so:spaces/mnzaki

export const appWrapReducer = (
  state = initialAppWrapState,
  action: AnyAction,
): AppWrapState => {
  let appWrapConfig, appWrapConfigsSet
  switch (action.type) {
    case APPWRAP_REGISTER_CONFIG:
      const idx = state.appWrapConfigsSet.indexOf(action.value)
      if (idx > -1 || Object.keys(action.value).length === 0) return state
      appWrapConfig = {...state.appWrapConfig, ...action.value }
      appWrapConfigsSet = [...state.appWrapConfigsSet, action.value]
      return { ...state, appWrapConfig, appWrapConfigsSet }
    case APPWRAP_UNREGISTER_CONFIG:
      const idx2 = state.appWrapConfigsSet.indexOf(action.value)
      if (idx2 < 0) return state
      appWrapConfigsSet = state.appWrapConfigsSet.filter(s => s !== action.value)
      appWrapConfig = appWrapConfigsSet.reduce((prev, cur) => ({...prev, ...cur}), initialAppWrapAttrs)
      return { ...state, appWrapConfig, appWrapConfigsSet }

    case APPWRAP_SHOW_LOADER:
      return {...state, appWrapConfig: {...state.appWrapConfig, loading: action.value} }
    case APPWRAP_SET_LOCKED:
      return { ...state, locked: action.value }

    case APPWRAP_SET_DISABLE_LOCK:
      return { ...state, disableLock: action.value }
    default:
      return state
  }
}
