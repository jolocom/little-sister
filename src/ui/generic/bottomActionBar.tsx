import { StyleSheet, View, TouchableOpacity } from 'react-native'
import { JolocomTheme } from 'src/styles/jolocom-theme'
import * as React from 'react'
import Icon from 'react-native-vector-icons/MaterialCommunityIcons'

const NAVIGATION_HEIGHT = 82
const NAVIGATION_CONTENT_HEIGHT = NAVIGATION_HEIGHT - 32

interface ActionBarProps {
  openScanner: () => void
}

const styles = StyleSheet.create({
  navigationContent: {
    height: NAVIGATION_CONTENT_HEIGHT,
    backgroundColor: '#fafafa',
    width: '100%',
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  navigationContentItem: {
    flex: 1,
    flexDirection: 'row',
    justifyContent: 'space-around',
    alignItems: 'center',
  },
  contentLeft: {
    marginRight: 36,
  },
  contentRight: {
    marginLeft: 36,
  },
  qrCodeButton: {
    position: 'absolute',
    bottom: 6,
    height: 72,
    width: 72,
    borderRadius: 35,
    backgroundColor: JolocomTheme.primaryColorPurple,
    alignItems: 'center',
    justifyContent: 'center',
    shadowColor: '#000',
    shadowOffset: { width: 0, height: 2 },
    shadowOpacity: 0.8,
    shadowRadius: 2,
    elevation: 8,
  },
})

export class BottomActionBar extends React.Component<ActionBarProps, {}> {
  render() {
    return (
      <View display={'flex'} alignItems={'center'}>
        <View style={styles.navigationContent}>
          <View style={[styles.navigationContentItem, styles.contentLeft]} />
          <View style={[styles.navigationContentItem, styles.contentRight]} />
        </View>

        <TouchableOpacity
          style={styles.qrCodeButton}
          onPress={this.props.openScanner}
        >
          <Icon size={30} name="qrcode-scan" color="white" />
        </TouchableOpacity>
      </View>
    )
  }
}
