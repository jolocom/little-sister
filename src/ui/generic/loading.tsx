import React from 'react'
import { Container, CenteredText } from 'src/ui/structure/'
import { Dimensions, Image, StyleSheet } from 'react-native'
import { JolocomTheme } from 'src/styles/jolocom-theme'

const image = require('src/resources/img/splashScreen.png')

interface Props {}

const styles = StyleSheet.create({
  loadingContainer: {
    backgroundColor: '#05050d',
  },
  loadingText: {
    position: 'absolute',
    bottom: '5%',
    fontSize: 12,
    fontFamily: JolocomTheme.contentFontFamily,
    color: JolocomTheme.primaryColorSand,
    opacity: 0.7,
  },
})

export const LoadingScreen: React.SFC<Props> = props => {
  const viewWidth: number = Dimensions.get('window').width
  const viewHeight: number = Dimensions.get('window').height

  return (
    <Container style={styles.loadingContainer}>
      <Image
        source={image}
        style={{
          bottom: '15%',
          width: viewWidth,
          height: viewHeight / 2,
        }}
      />
      <CenteredText style={styles.loadingText} msg={'POWERED BY JOLOCOM'} />
    </Container>
  )
}
