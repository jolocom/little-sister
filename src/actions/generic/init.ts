import { Linking } from 'react-native'
import SplashScreen from 'react-native-splash-screen'
import { AppError, ErrorCode } from '../../lib/errors'
import I18n from '../../locales/i18n'
import { routeList } from '../../routeList'
import { navigationActions, accountActions, genericActions } from '../../actions'
import { ThunkAction } from '../../store'
import { withLoading, withErrorScreen } from '../modifiers'
import { showErrorScreen } from '.'

export const initApp: ThunkAction = async (
  dispatch,
  getState,
  backendMiddleware,
) => {
  try {
    const storedSettings = await backendMiddleware.storage.get.settingsObject()

    // locale setup
    if (storedSettings.locale) I18n.locale = storedSettings.locale.selected
    else storedSettings.locale = I18n.locale

    storedSettings.locale = I18n.locale

    await dispatch(loadSettings(storedSettings))

    const ret = await dispatch(accountActions.checkIdentityExists)

    // FIXME what happens if no identity and this is a deeplink?
    // navigationActions.handleDeepLink throws NoWallet
    // need to improve this UX

    // FIXME: get rid of these after setting up deepLinking properly using
    // react-navigation
    const handleDeepLink = (url: string) =>
      dispatch(
        withLoading(withErrorScreen(navigationActions.handleDeepLink(url))),
      )
    Linking.addEventListener('url', event => handleDeepLink(event.url))
    const deepLinkUrl = await Linking.getInitialURL()
    if (deepLinkUrl) {
      await handleDeepLink(deepLinkUrl)
    }

    await dispatch(genericActions.lockApp())
    return ret
  } catch (e) {
    return dispatch(
      showErrorScreen(
        new AppError(ErrorCode.WalletInitFailed, e, routeList.Landing),
      ),
    )
  } finally {
    SplashScreen.hide()
  }
}

const loadSettings = (settings: { [key: string]: any }) => ({
  type: 'LOAD_SETTINGS',
  value: settings,
})
