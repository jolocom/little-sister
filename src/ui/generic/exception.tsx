import React from 'react'
import { connect } from 'react-redux'
import { navigationActions } from 'src/actions/'
import {
  Text,
  StyleSheet,
  View,
  Image,
  BackHandler,
  StatusBar,
} from 'react-native'
import I18n from 'src/locales/i18n'
import { errorTitleMessages, AppError } from '../../lib/errors'
import { getRandomStringFromArray } from 'src/utils/getRandomStringFromArray'
import strings from 'src/locales/strings'
import { ThunkDispatch } from '../../store'
import { NavigationInjectedProps } from 'react-navigation'
import { Colors, Spacing, Typography } from 'src/styles'
import { JolocomButton } from '../structure'
import { routeList } from '../../routeList'
const errorImage = require('src/resources/img/error_image.png')

interface Props
  extends ReturnType<typeof mapDispatchToProps>,
    ReturnType<typeof mapStateToProps>,
    NavigationInjectedProps {
  errorTitle?: string
}

interface State {
  reportSent: boolean
}

const styles = StyleSheet.create({
  container: {
    backgroundColor: Colors.blackMain,
    flex: 1,
    justifyContent: 'space-around',
    alignItems: 'center',
    padding: '5%',
  },
  upperContainer: {
    justifyContent: 'space-around',
    alignItems: 'center',
    marginTop: Spacing['4XL'],
  },
  textBlock: {
    marginTop: Spacing.LG,
    paddingHorizontal: Spacing.MD,
    alignItems: 'center',
  },
  errorTextHeader: {
    ...Typography.mainText,
    textAlign: 'center',
    color: Colors.sandLight,
  },
  errorText: {
    ...Typography.subMainText,
    textAlign: 'center',
    color: Colors.sandLight,
    marginTop: Spacing.MD,
  },
  buttonBlock: {
    marginTop: Spacing.LG,
    justifyContent: 'space-evenly',
    height: 150,
    width: '100%',
  },
})

export class ExceptionComponent extends React.Component<Props, State> {
  private onBackButtonPressAndroid = (): boolean => {
    this.handleTapBack()
    return true
  }
  public componentDidMount(): void {
    BackHandler.addEventListener(
      'hardwareBackPress',
      this.onBackButtonPressAndroid,
    )
  }

  public componentWillUnmount(): void {
    BackHandler.removeEventListener(
      'hardwareBackPress',
      this.onBackButtonPressAndroid,
    )
  }

  private handleTapBack = (): void => {
    const { navigation } = this.props
    if (navigation) {
      this.props.navigateBack(navigation?.state?.params?.returnTo)
    }
  }

  private getError = (): AppError | Error | undefined => {
    return this.props.navigation?.state?.params?.error
  }

  public render(): JSX.Element | null {
    // TODO: display error code
    const err = this.getError()
    const origError = err instanceof AppError && err.origError
    const origErrorMessage = origError && origError.message
    const errorTitle =
      this.props.errorTitle || getRandomStringFromArray(errorTitleMessages)
    let errorText = err
      ? err.message
      : strings.THERE_WAS_AN_ERROR_WITH_YOUR_REQUEST
    errorText = I18n.t(errorText) + '.'
    console.error(origError || err)

    return (
      <View style={styles.container}>
        <StatusBar barStyle="light-content" />
        <View style={styles.upperContainer}>
          <Image source={errorImage} style={{ width: 160, height: 160 }} />
          <View style={styles.textBlock}>
            <Text style={styles.errorTextHeader}>
              {I18n.t(errorTitle) + '.'}
            </Text>
            <Text numberOfLines={5} style={styles.errorText}>
              {errorText}
            </Text>
            {origErrorMessage && (
              <Text numberOfLines={5} style={styles.errorText}>
                {origErrorMessage}
              </Text>
            )}
          </View>
        </View>
        <View style={styles.buttonBlock}>
          {err && (
            <JolocomButton
              onPress={() => this.props.navigateReporting(err)}
              text={I18n.t(strings.SEND_ERROR_REPORT)}
            />
          )}
          <JolocomButton
            transparent
            onPress={this.handleTapBack}
            text={I18n.t(strings.GO_BACK)}
          />
        </View>
      </View>
    )
  }
}

const mapStateToProps = (): {} => ({})

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  navigateBack: (routeName: string) =>
    dispatch(navigationActions.navigate({ routeName })),
  navigateReporting: (error: AppError | Error | undefined) =>
    dispatch(
      navigationActions.navigate({
        routeName: routeList.ErrorReporting,
        params: { error },
      }),
    ),
})

export const Exception = connect(
  mapStateToProps,
  mapDispatchToProps,
)(ExceptionComponent)
