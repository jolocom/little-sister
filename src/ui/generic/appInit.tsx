import * as React from 'react'
import { connect } from 'react-redux'

import { ThunkDispatch } from '../../store'
import { Image } from 'react-native'
import { withErrorHandler } from '../../actions/modifiers'
import { Wrapper } from '../structure'
import { AppError, ErrorCode } from '../../lib/errors'
import { showErrorScreen } from '../../actions/generic'
import { initApp } from 'src/actions/generic/init'

interface Props extends ReturnType<typeof mapDispatchToProps> {}

export class AppInitContainer extends React.Component<Props> {
  constructor(props: Props) {
    super(props)
    this.props.doAppInit()
  }

  render() {
    return (
      <Wrapper dark withoutSafeArea centered>
        <Image
          source={require('src/resources/img/splashIcons/joloLogoIcon.png')}
        />
      </Wrapper>
    )
  }
}

const mapDispatchToProps = (dispatch: ThunkDispatch) => ({
  doAppInit: () =>
    dispatch(
      withErrorHandler(
        showErrorScreen,
        (err: Error) => new AppError(ErrorCode.AppInitFailed, err),
      )(initApp),
    ),
})

export const AppInit = connect(null, mapDispatchToProps)(AppInitContainer)
