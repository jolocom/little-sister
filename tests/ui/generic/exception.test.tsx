import React from 'react'
import { shallow } from 'enzyme'
import { ExceptionComponent } from 'src/ui/generic/exception'
import { StyleSheet } from 'react-native'

describe('Exception screen component', () => {
  it('Renders correctly', () => {
    const props = {
      navigation: {
        state: {
          params: {
            flag: 'default',
          },
        },
      },
    }

    const rendered = shallow(<ExceptionComponent {...props} />)
    expect(rendered).toMatchSnapshot()
  })

  it('Renders correctly when no error object is provided', () => {
    const props = {
      navigation: {
        state: {
          params: {
            errorMessage: '',
            stackTrace: undefined,
          },
        },
      },
    }

    const rendered = shallow(<ExceptionComponent {...props} />)
    expect(rendered).toMatchSnapshot()
  })
})
