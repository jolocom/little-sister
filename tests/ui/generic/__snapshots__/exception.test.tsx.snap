// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exception screen component Renders correctly 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(5, 5, 13)",
      "flex": 1,
      "justifyContent": "space-around",
      "padding": "5%",
    }
  }
>
  <StatusBar
    animated={false}
    barStyle="light-content"
    showHideTransition="fade"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "space-around",
        "marginTop": 80,
      }
    }
  >
    <Image
      source={
        Object {
          "process": [Function],
        }
      }
      style={
        Object {
          "height": 160,
          "width": 160,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "marginTop": 24,
          "paddingHorizontal": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(255, 239, 223)",
            "fontFamily": "TT Commons",
            "fontSize": 30,
            "textAlign": "center",
          }
        }
      >
        Uh oh.
      </Text>
      <Text
        numberOfLines={5}
        style={
          Object {
            "color": "rgb(255, 239, 223)",
            "fontFamily": "TT Commons",
            "fontSize": 20,
            "marginTop": 16,
            "textAlign": "center",
          }
        }
      >
        Unknown Error.
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "height": 150,
        "justifyContent": "space-evenly",
        "marginTop": 24,
        "width": "100%",
      }
    }
  >
    <JolocomButton
      onPress={[Function]}
      text="Send error report"
    />
    <JolocomButton
      onPress={[Function]}
      text="Go back"
      transparent={true}
    />
  </View>
</View>
`;

exports[`Exception screen component Renders correctly when no error object is provided 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "rgb(5, 5, 13)",
      "flex": 1,
      "justifyContent": "space-around",
      "padding": "5%",
    }
  }
>
  <StatusBar
    animated={false}
    barStyle="light-content"
    showHideTransition="fade"
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "justifyContent": "space-around",
        "marginTop": 80,
      }
    }
  >
    <Image
      source={
        Object {
          "process": [Function],
        }
      }
      style={
        Object {
          "height": 160,
          "width": 160,
        }
      }
    />
    <View
      style={
        Object {
          "alignItems": "center",
          "marginTop": 24,
          "paddingHorizontal": 16,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "rgb(255, 239, 223)",
            "fontFamily": "TT Commons",
            "fontSize": 30,
            "textAlign": "center",
          }
        }
      >
        Uh oh.
      </Text>
      <Text
        numberOfLines={5}
        style={
          Object {
            "color": "rgb(255, 239, 223)",
            "fontFamily": "TT Commons",
            "fontSize": 20,
            "marginTop": 16,
            "textAlign": "center",
          }
        }
      >
        There was an error with your request.
      </Text>
    </View>
  </View>
  <View
    style={
      Object {
        "height": 150,
        "justifyContent": "space-evenly",
        "marginTop": 24,
        "width": "100%",
      }
    }
  >
    <JolocomButton
      onPress={[Function]}
      text="Go back"
      transparent={true}
    />
  </View>
</View>
`;
